apps/prairielearn/python/zygote.py

```diff
@@ -31,15 +31,29 @@
 
 drop_privileges = os.environ.get("DROP_PRIVILEGES", False)
 
-# If we're configured to drop privileges, matplotlib won't be able to write to
-# its default config/cache dir. To keep it fast, we'll create a cache dir for
-# it and provide it with config via the `MPLCONFIGDIR` environment variable.
+# If we're configured to drop privileges (that is, if we're running in a
+# Docker container), various tools like matplotlib and fontconfig will be
+# unable to write to their default config/cache directories. This is because
+# the `$HOME` environment variable still points to `/root`, which is not
+# writable by the `executor` user.
+#
+# To work around this, we'll set `$XDG_CONFIG_HOME` and `$XDG_CACHE_HOME` to
+# directories created in `/tmp` that are world-writable. matplotlib and
+# fontconfig should respect these environment variables; other tools should as
+# well. If they don't, special cases can be added below.
 if drop_privileges:
-    config_dir_path = "/tmp/matplotlib"
+    config_home_path = "/tmp/xdg_config"
+    cache_home_path = "/tmp/xdg_cache"
```

Comment:
Will these be shared among zygotes? Is this a problem?

Reply:
There is only one Python process executing per container; this is not a problem.

Reply:
I remember in #7298 this was an issue. Can you clarify why this is not a problem here?

Reply:
That is only an issue in situations where we _do_ in fact run multiple zygotes that share the same underlying filesystem, which is true in development and in most self-hosted situations. 

The undocumented assumption is that if `DROP_PRIVILEGES` is set, we're running in a container, which currently holds true for all the ways that we run PL. In production, we run _exactly_ one zygote per container (and thus per filesystem).